kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -481,9 +478,14 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
                                 // or we are adding a child to a coroutine that is not completing yet
                                 if (rootCause == null || handler.isHandlerOf<ChildHandleNode>() && !state.isCompleting) {
                                     // Note: add node the list while holding lock on state (make sure it cannot change)
-                                    val node = makeNode(handler, onCancelling)
-                                    val res = addLast(state, list, node).also { nodeAdded = node }
-                                    if (!res) return@loopOnState // retry
+                                    val node = nodeAdded ?: makeNode(handler, onCancelling).also { nodeAdded = it }
+                                    if (!addLast(state, list, node)) {
+                                        // parent failed to add ChildCompletion node to the child, the child became Completed
```

Comment:
I've moved that `ChildCompletion` logic from `addLast` impl here.

Consider this scenario:
The Parent is `Finishing` and starts to wait for a child to complete. It tries to add `ChildCompletion` node to the child:
 
if it succeeds and `ChildCompletion` is added before the child changes its state to `Completed`, then the parent just waits for the child

if the child has already completed, and there are no other children left to wait for, then the parent starts it's final transition to `Completed` state.

But as adding to the list is not atomic with the state change now, the child may change it's state to `Completed` and see the list with `ChildCompletion` added before this node is disposed, invoke `ChildCompletion` and start parent's terminal state transition to `Completed` state (`finalizeFinishingState`). While the parent reports, that the child has already completed, no one to wait for and starts it's transition to the `Completed` state itself (concurrently with the child). 

So, I check here whether `ChildCompletion` was already invoked and if so, the parent should wait for the child completion.

Reply:
That's a really useful comment, thanks for the explanation!

I would appreciated if you could embed it into the code comment in a slightly more condensed form